### ISSUE TITLE
Implement BitTorrent tracker

### ISSUE THREAD

**Issue by radumarias** [radumarias/rfs#4]
https://en.m.wikipedia.org/wiki/Bencode
Bencode (pronounced like Bee-encode) is the encoding used by the peer-to-peer file sharing system BitTorrent for storing and transmitting loosely structured data.

https://www.bittorrent.org/beps/bep_0003.html#trackers
https://chatgpt.com/share/e8b32340-1eb7-4865-8ae7-b14a1977e56c

**Comment by bnchi** [comment]
I was thinking what happen if a user decide to do write/update on a file are we going to allow it ? 

Imagine this scenario : 

1. User A decide to update file x
2. file exist on a system with multiple replicas

Letting the user update a file will result in us having to find all replicas removing them and re-applying whatever changes. I think a naive solution like this is expensive and might lead to issues if 2 users are editing a file at the same.

I propose that files are immutable on the system, that's once a file is replicated and in the system it can't be modified. However the user is free to pull a file update it and re-upload again this will result in a more storage but might eliminate any issues with us having to deal with synchronization on the system. 

What are your thoughts ?

**Comment by bnchi** [comment]
Just a thought as well if we know the topology of our system do we need to find peers to download the chunk from ?

If the user define how many nodes and the topology of the system that's we have x amount of nodes in the cluster do we really need to implement a peer exchange protocol.

Let me explain a little bit what I mean here : 

Imagine we have cluster with nodes numbered n1 .. n3

At start each node might not know who it's neighbor is that is n1 doesn't know n2,n3 doesn't know n1 etc .. 

What we want is once a system is bootstrapped to be able to let each node know the topology of the system, consider this simple JSON representation that describe a topology : 

```
{
    "n1": ["n2", "n3"],
    "n2": ["n1"],
    "n3": ["n1"]
}
```

Imagine this a graph with each node connected to the other with an edge.

What we want really is to broadcast this topology to all the nodes on the server, once a new node comes into the system same thing broadcast and update.

With this now each node in the system have a in-memory representation of how to communicate with each other thus allowing us to find each chunk in the system.

I might be missing something here let me know what do you think ?

**Comment by radumarias** [comment]
>If the user define how many nodes and the topology of the system that's we have x amount of nodes in the cluster do we really need to implement a peer exchange protocol.

As I understand PEX sends list of peers for the file and also what each peer what shards they have and what chunks from the sharks (like seeder)

The network topology I assume could be also synched with Raft, just we can use PEX for the above shards and chunks metadata

But do I get it wrong on what PEX does?

**Comment by bnchi** [comment]
Let me try to describe the flow of the system and with that we can probably have a general idea on what we're actually building for this issue.

I think the design is starting to shape in my head now as the following : 

- Once a user add a new file to the system these steps happen : 
 - The file is divided into chunks ideally it's not more than 256 KB per chunk 
 - SHA is computed for each piece
 - Now in the tikv store we store the metadata about the file, the metadata include the list of hashed pieces for a file, the size of the file,  size of each piece, and the original name of the file.
 - We're then going to distributed the files into the system without replication for starter to get the protocol working 

Now in the above step we can decide on multiple strategies : 

1. Storing a DHT of all the nodes that have that piece and then deciding on which node is the closest to the user to allow them to connect into and download the piece. This will allow us probably to build a plan with nodes that are local to the end user thus faster download time.
2. Using PEX  to propagate through the system and find a node that has the piece, from my understating of PEX you need a set of initial peers stored in a torrent/tracker/DHT whatever and then the protocol does it's magic by finding other peers that you can connect to and download a piece.

Now I favor option 1 for simplicity because it's clear, but we will probably have to think how to handle failure that's during the time of downloading a piece a server might fail thus trying to find another one that has that exact same piece.

Downloading a piece flow that's the client work : 

- Once a user decide to get a file from the system : 
 - We match against our tikv store to get the pieces details 
 - now we can try to build a plan by finding all the nodes that are close to a user and contain that piece.
 - Now we will connect to each peer and try to download the file

**Comment by radumarias** [comment]
I would still try with PEX also as seems more reliable and more scalable.
Storing initial metadata in DHT seems better that torrent file as it's more scalable and fault tolerant

>The file is divided into chunks, ideally not more than 256 KB per chunk.

The actual shards (splits of the original file) I would imagine 512MB or 64MB and then from torrent POV each of these shards is seen as a file where yes chunk should be in the order of hundreds of kB

**Comment by bnchi** [comment]
The idea here is that with a DHT we can also maintain a list for replicas that's : 

the key is the piece hash and the value is a list of all the nodes that includes those replicas we still need to think about replication but that's how I imagine we're going to model the replication of a piece  

I still don't get how are we going to be fault tolerance with PEX.

**Comment by bnchi** [comment]
I think this wiki describes PEX and DHT in a much better way than Wikipedia give it a read and let me know what you think @radumarias 


http://wiki.bitcomet.com/Peers,_seeds,_torrent,_tracker,_DHT,_Peer_Exchange_%28PEX%29,_Magnet_Links#Peer_Exchange_.28PEX.29

**Comment by radumarias** [comment]
I agree we don't need PEX but DHT. I think it's more scalable to use DHT instead of a tracker as it eliminates the single point of failure.

Will we need to create some service that acts like DHT and reads metadata from tikv?

**Comment by bnchi** [comment]
https://dl.acm.org/doi/10.5555/646334.687801 the Kademlia paper if we decided to go with it ..
